### ISSUE TITLE
Use GPG2, Import GPG keys using HKP over HTTPS

### ISSUE THREAD

**Issue by cpu** [StreisandEffect/streisand#963]
Presently Streisand uses `x-hkp://pool.sks-keyservers.net` globally as the keyserver address. We also use the default system`gpg` (version 1.4). 

This means we fetch keys over port 11371 (The default HKP port) using plaintext HTTP.  From the perspective of the integrity of the received key it isn't detrimental to fetch keys over HTTP. The key is signed and we [verify the fingerprint of the key received matches the full fingerprint expected](https://github.com/jlund/streisand/blob/c58d33776cf19570486858d636aad1d8c1c09d2f/playbooks/roles/download-and-verify/tasks/main.yml#L16).

It would be slightly better from a privacy perspective if we used HKPS, the HKP protocol over HTTPS.

The `sks-keyservers.net` pool includes `hkps.pool.sks-keyservers.net` for this. According to the sks-keyservers pool overview this domain listens on port 443:
> This is a pool containing only servers available using hkps. Regular A and AAAA and SRV records are included for port 443 servers, and a lookup is performed for _pgpkey-https._tcp on the individual servers to determine if a hkps enabled service is listening on another port. At this point, however, servers not running on port 443 are not included.

Using HKPS with the sks-keyservers pool requires verifying the server's leaf certificate using a custom CA certificate made available by the sks-keyservers pool. This should be installed somewhere other than `/etc/ssl/certs` since we don't want it to be a trusted CA system-wide. Instead we should add a `--keyserver-options ca-cert-file=/etc/whatever/path/to/ca.pem` line to the GPG invocation. 

I've experimented with this a little bit [in a branch on my fork](https://github.com/cpu/streisand/commit/0a560b9879f2686aa0e187120c682ee616133799). In practice I wasn't able to get this to work with `gpg` and had to use `gpg2` (Which we don't install by default). It also threw a deprecation warning about using `--keyserver-options ca-cert-file` vs setting a `~/.gnupg/dirmgr.conf` setting.

I've put this on hold for now (too many other things on the go) but hope to return to it shortly so I'll leave it assigned to myself for now.

**Comment by affectiosus** [comment]
@cpu hey, any news regarding this issue? It's been over two months without any updates :(

**Comment by cpu** [comment]
Hi @affectiosus - Unfortunately my last update about putting it on hold is still accurate.

Thanks for the bump - knowing you're waiting on this is a good motivator to try and pick it up again. I'll update this ticket when I've started in earnest. Are you looking at this for the same reasons as were expressed in #960, the restrictive egress firewall?

**Comment by affectiosus** [comment]
@cpu yes, seems like the provider I'm currently stuck with has some never mentioned hidden restrictions.

**Comment by nikksno** [comment]
+1. Let me know if I can help. Thanks again!

**Comment by nikksno** [comment]
Keep in mind that if this were to become too complicated we can simply leave connections unencrypted but at least use port 80 by replacing `{{ gpg_key_server_address }}` with `{{ gpg_key_server_address }}:80` in `streisand/playbooks/roles/tor-bridge/tasks/mirror-common.yml` and `streisand/playbooks/roles/download-and-verify/tasks/main.yml`. It does not solve the issue, but it's a start to at least get around blocked ports on filtered networks.

**Comment by cpu** [comment]
I picked this ticket up again the other night. I have most of the work for the HKPS component finished. I'm working on doing some general cleanup of the GPG related portions of Streisand and hope to have all of this bundled into a PR by the end of the weekend.

**Comment by cpu** [comment]
@nikksno I think #1231 will meet your needs, want to give it a :mag: ?

**Comment by cpu** [comment]
@nikksno @affectiosus The HKPS rework landed in master so I'm going to close this issue. I'm interested in hearing how it works for you folks.

**Comment by nikksno** [comment]
Awesome work thank you so much, I'll try it out ASAP.
